File: JMP.Enterprises/backend/JMP.Enterprises.Api/Models/StatementOfAccount.cs
namespace JMP.Enterprises.Api.Models;

/// <summary>
/// Represents a monthly Statement of Account (SOA) for long-stay rental agreements.
/// Maps to "StatementsOfAccount" table in SQL Server.
/// </summary>
public class StatementOfAccount
{
    public int StatementOfAccountId { get; set; }

    /// <summary>
    /// Unique SOA reference number (e.g., JMP-SOA-202609-0001)
    /// </summary>
    public string SoaNumber { get; set; } = string.Empty;

    public int ReservationId { get; set; }
    public Reservation? Reservation { get; set; }

    public int RentalAgreementId { get; set; }
    public RentalAgreement? RentalAgreement { get; set; }

    public DateTime BillingPeriodStart { get; set; }

    public DateTime BillingPeriodEnd { get; set; }

    public DateTime DueDate { get; set; }

    public DateTime IssueDate { get; set; } = DateTime.Now;

    // Monthly Rent Charge
    public decimal MonthlyRentAmount { get; set; }

    // Electricity Submeter Details
    public decimal PreviousElectricityReading { get; set; } = 0;

    public decimal PresentElectricityReading { get; set; } = 0;

    public decimal ElectricityConsumptionKwh { get; set; } = 0;

    public decimal ElectricityRatePerKwh { get; set; } = 0;

    public decimal ElectricityAmount { get; set; } = 0;

    // Other Utilities
    public decimal WaterAmount { get; set; } = 0;

    public decimal InternetAmount { get; set; } = 0;

    public decimal AdditionalCharges { get; set; } = 0;

    public string? AdditionalChargesDescription { get; set; }

    // Balances & Calculations
    public decimal PreviousBalance { get; set; } = 0;

    public decimal TotalAmountDue { get; set; } = 0;

    public decimal AmountPaid { get; set; } = 0;

    public decimal BalanceRemaining { get; set; } = 0;

    /// <summary>
    /// Pending | PartiallyPaid | Paid | Overdue
    /// </summary>
    public string Status { get; set; } = "Pending";

    public string? Notes { get; set; }

    public DateTime CreatedDate { get; set; } = DateTime.Now;

    public DateTime? UpdatedDate { get; set; }
}

File: JMP.Enterprises/frontend/src/types/statementOfAccount.ts
export interface StatementOfAccount {
  statementOfAccountId: number;
  soaNumber: string;
  reservationId: number;
  rentalAgreementId: number;
  agreementNumber: string;
  tenantName: string;
  tenantCompanyName?: string;
  propertyName: string;
  propertyCode: string;
  billingPeriodStart: string;
  billingPeriodEnd: string;
  dueDate: string;
  issueDate: string;
  monthlyRentAmount: number;
  previousElectricityReading: number;
  presentElectricityReading: number;
  electricityConsumptionKwh: number;
  electricityRatePerKwh: number;
  electricityAmount: number;
  waterAmount: number;
  internetAmount: number;
  additionalCharges: number;
  additionalChargesDescription?: string;
  previousBalance: number;
  totalAmountDue: number;
  amountPaid: number;
  balanceRemaining: number;
  status: 'Pending' | 'PartiallyPaid' | 'Paid' | 'Overdue';
  notes?: string;
  createdDate: string;
  updatedDate?: string;
}

export interface CreateStatementOfAccountDto {
  reservationId: number;
  rentalAgreementId: number;
  billingPeriodStart: string;
  billingPeriodEnd: string;
  dueDate: string;
  monthlyRentAmount: number;
  previousElectricityReading: number;
  presentElectricityReading: number;
  electricityRatePerKwh: number;
  waterAmount: number;
  internetAmount: number;
  additionalCharges: number;
  additionalChargesDescription?: string;
  previousBalance: number;
  notes?: string;
}

export interface UpdateStatementOfAccountDto {
  billingPeriodStart: string;
  billingPeriodEnd: string;
  dueDate: string;
  monthlyRentAmount: number;
  previousElectricityReading: number;
  presentElectricityReading: number;
  electricityRatePerKwh: number;
  waterAmount: number;
  internetAmount: number;
  additionalCharges: number;
  additionalChargesDescription?: string;
  previousBalance: number;
  status: 'Pending' | 'PartiallyPaid' | 'Paid' | 'Overdue';
  notes?: string;
}

export interface RecordSoaPaymentDto {
  paymentAmount: number;
  paymentMethod: string;
  referenceNumber?: string;
  notes?: string;
}
